werkit/compute/graph/_dependency_graph.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import annotations
+import inspect
+import numbers
+import typing as t
+from typing_extensions import Literal, TypedDict
+from ._built_in_type import (
+    BuiltInValueType,
+    coerce_value_to_builtin_type,
+    is_built_in_value_type,
+)
+from ._custom_type import CustomType, JSONType
+
+
+AnyValueType = t.Union[BuiltInValueType, t.Type[CustomType]]
+
+
+class BaseNode:
+    def __init__(self, value_type: AnyValueType):
+        if is_built_in_value_type(value_type) or (
+            inspect.isclass(value_type) and issubclass(value_type, CustomType)
+        ):
+            self.value_type = value_type
+        else:
+            raise ValueError(
+                "Expected value type to be bool, int, float, str, or a subclass of CustomType"
+            )
+
+    @property
+    def value_type_is_built_in(self) -> bool:
+        return is_built_in_value_type(self.value_type)
+
+    @property
+    def value_type_name(self) -> str:
+        if self.value_type_is_built_in:
```

Comment:
Is there a reason why we want special names for the built-in value types? e.g. "Boolean" as opposed to "bool". I ask, because the built-in types also support the `.__name__` property, so this code in value_type_name could be simplified. And we will need to apply an alias/mapping on the typescript side anyway, e.g. from "Boolean" or "bool" to Typescript's "boolean". Why should we give these built-in types a special name, if we will need to apply a mapping in the consuming code anyway?

Reply:
This and [your other comment](https://github.com/metabolize/werkit/pull/228#discussion_r824912261) are part of a design question around which built-in types the compute graph should support, and what they should be called. Since the compute graph is deliberately a polyglot JSON interface, I would prefer we choose what makes sense for the interface (rather than what is easiest to implement in Python, as it's trivially easy to map them).

On the broader question of number types: whereas GraphQL, Python, and JSON Schema have separate Int and Float types, JSON and JavaScript do not. I can't think of a reason we need to support separate int / float types at the interface layer, so I think it makes sense to have a single number type to make the interface smaller. Suppose I'm open to being convinced on this. Think the main challenge is figuring out how to map an integer through 1schema / ts-json-schema-generator (not sure it's supported yet).

Then the question of consistent naming. For a polyglot system I probably wouldn't choose the Python-specific names. We could use the TypeScript names, which could make sense since we're using 1schema (TS + JSON Schema) on both ends as a sort of common interface. `'string'`, `'number'`, `'boolean'`. They're the same in JSON Schema, which is nice.

Might make sense to discuss offline.

Reply:
> Then the question of consistent naming. For a polyglot system I probably wouldn't choose the Python-specific names. We could use the TypeScript names, which could make sense since we're using 1schema (TS + JSON Schema) on both ends as a sort of common interface. 'string', 'number', 'boolean'. They're the same in JSON Schema, which is nice.

Using typescript type names for this makes sense to me, as typescript and JSON schema are already the language we are using to define APIs for consumption across language boundaries.